src/types/CanonicalField.ts

```diff
@@ -0,0 +1,57 @@
+import { ajv } from '../ajv';
+import type { JSONSchemaType } from 'ajv';
+
+export interface CanonicalField {
+  id: number;
```

Comment:
So this is great for `CanonicalField` but I anticipate with most types we are going to need (at least) two options: one for reading and one for writing. When posting data, we cannot post an ID, we want that auto-generated, and probably the same for created datetime, although as noted above, for test purposes it seems like we want to specify it.

Reply:
Yes!  So in those cases I think we're going to want to use the TypeScript [utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html) which lets us define a full interface and then modify certain fields to be required / optional based on the context.

So we define the full type in one place, and then we define the subsets directly in the various handlers.

All of that said, I don't know *exactly* how the ajv / validation side of that pattern will look, but I do think that path at least means we don't have to define a bunch of sub-types in advance / before they're needed.

Reply:
Right, well in some contexts we want to require the id (reading) and in other contexts we want to require the id to not be set (writing), so it's not exactly optional, either. My approach of the following (which I see your similar comments on) was my first attempt but probably fails to achieve what I'm describing anyway:
```TypeScript
export interface OpportunityMinimal {
  title: string;
}

export interface Opportunity extends OpportunityMinimal {
  id: bigint;
  created_at: Date;
}

// export type Opportunity = OpportunityMinimal | OpportunityMaximal;
```
Actually, my first attempt had that second thing being `OpportunityMaximal extends OpportunityMinimal` and then the third thing being the union. That might have actually described what I wanted but the linter said "no dice, bad style."

Reply:
> in some contexts we want to require the id (reading) and in other contexts we want to require the id to not be set

Yes!  So what I'm saying is that *in those contexts* we would use the utility types to create context specific types.

It might be that we want to create our own utility class For example: https://bobbyhadz.com/blog/typescript-make-single-property-optional

Basically instead of defining several versions of the type types outside of the specific context we would do something like this *when defining the parameter*

for instance....

```
import { MakeRequired } from './types
interface CanonicalFieldWithRequiredId = MakeRequired<CanonicalField, 'id'>

function updateCanonicalField(
  canonicalField: CanonicalFieldWithRequiredId
)
```

I think that probably the core types should be based on the schema -- so if something is nullable in DB it is nullable in the type, but anything schema-required is required in type (and in create contexts we make it nullable via a utility type).

So: if a given endpoint doesn't require a field then we would define a new type just for that particular endpoint.

Reply:
OK very nice. Thanks for your patience and the link, and the explanation over in https://github.com/PhilanthropyDataCommons/service/pull/49#discussion_r927007792 too. This helped, and the mention of `Omit` specifically resolves the thing I was stuck on (but was too distracted to look up in the docs that this was possible).

So tying it back to my example, the suggestion is to maximally define `Opportunity` (kind of like my `OpportunityMaximal` and then subtract as-needed in the context needed using utility types. My attempt at understanding it more clearly:

```TypeScript
interface Opportunity {
    id: number;
    title: string;
    createdAt: Date;
}

const oppFull: Opportunity = {
    id: 1,
    title: 'chicken',
    createdAt: new Date(),
}

console.log(`${oppFull}`);

function printOpportunity(minOpp: Omit<Opportunity, 'createdAt' | 'id'>) {
    console.log(`minOpp: ${minOpp}`);

    // Property does not exist, cool:
    // console.log(`minOpp.id: ${minOpp.id}`);
    console.log(`minOpp.title: ${minOpp.title}`);

    // Property createdAt does not exist, cool:
    //console.log(`minOpp.createdAt: ${minOpp.createdAt}`);
}

printOpportunity(oppFull);
```

With utility types it looks like we could do it in a subtractive way (kind of like milling) or additive way (kind of like 3D printing) but I think I see your point now. We pick one definition, minimal or maximal as "the thing in general", and then we use the utility types in context. I think I got it, yes?

Reply:
We might even be able to genericize it somewhat with like `Omit<T, 'createdAt' | 'id'>` or something where T is generic type. Idk.

Reply:
Yes, that seems to work fine, actually:
```TypeScript

interface CanonicalField {
    id: number;
    description: string;
    createdAt: Date;
}

type Minimal<T> = Omit<T, 'createdAt' | 'id'>;

function printMinimalCanonicalField(minField: Minimal<CanonicalField>) {
    console.log(`minField.description: ${minField.description}`);
}

function printMinimalOpportunity(minOpp: Minimal<Opportunity>) {
    console.log(`minField.description: ${minOpp.title}`);
}

```

Reply:
I *think* that at this point this can be marked as resolved -- id is required at the schema level and we can `Omit` it when it's not needed in a given context.  If that's not an accurate takeaway please un-resolve!